src/contrib/mlir/compiler.cpp

```diff
@@ -0,0 +1,439 @@
+
+#include "dialect/dialect.hpp"
+#include "dialect/ops.hpp"
+#include "dialect/type.hpp"
+#include "lowerer.hpp"
+#include "ngraph/descriptor/tensor.hpp"
+#include "ngraph/graph_util.hpp"
+#include "ngraph/node.hpp"
+#include "ngraph/op/add.hpp"
+#include "ngraph/op/dot.hpp"
+#include "ngraph/op/experimental/compiled_kernel.hpp"
+#include "ngraph/type/element_type.hpp"
+
+#include <llvm/ADT/STLExtras.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Support/ErrorOr.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <llvm/Support/SourceMgr.h>
+#include <llvm/Support/TargetSelect.h>
+#include <mlir/ExecutionEngine/ExecutionEngine.h>
+#include <mlir/ExecutionEngine/MemRefUtils.h>
+#include <mlir/ExecutionEngine/OptUtils.h>
+#include <mlir/LLVMIR/LLVMDialect.h>
+#include <mlir/LLVMIR/Transforms.h>
+#include <mlir/Pass/PassManager.h>
+#include <mlir/Target/LLVMIR.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/Transforms/Passes.h>
+
+#include <memory>
+#include <mutex>
+
+using llvm::SmallVector;
+using llvm::StringRef;
+using llvm::make_unique;
+using namespace ngraph::runtime::ngmlir;
+
+#define COMPILE_OP_DECL(op_name)                                                                   \
+    create_op<op_name>(MLIRCompiler & compiler, const ngraph::Node* ng_node)
+
+MLIRCompiler::MLIRCompiler(const ngraph::op::CompiledKernel* compiled_kernel,
+                           const std::vector<void*>& external_tensors)
+    : m_compiled_kernel(compiled_kernel)
+    , m_external_tensors(external_tensors)
+{
+    NGRAPH_ASSERT((m_compiled_kernel->get_arguments().size() +
+                   m_compiled_kernel->get_kernel_outputs().size()) == external_tensors.size())
+        << "Number of arguments and outputs doesn't match number of tensors";
+}
+
+void MLIRCompiler::init_mlir()
+{
+    // Mutex to safely initialize MLIR.
+    static std::mutex mlir_init_mutex;
+    static bool initialized = false;
+
+    std::unique_lock<std::mutex> lock(mlir_init_mutex);
+
+    if (!initialized)
+    {
+        mlir::registerDialect<mlir::NGDialect>();
+        // Register any LLVM command line options
+        llvm::cl::ParseEnvironmentOptions("ngraph", "MLIR_LLVM_OPTIONS", "");
+        initialized = true;
+    }
+}
+
+void MLIRCompiler::compile_and_run()
+{
+    build_ng_dialect_module();
+    lower_ng_dialect();
+    optimize();
+    bind_arguments();
+    execute();
+    cleanup();
+}
+
+// Creates an MLIR module and function with nGraph dialect ops from the input CompiledKernel.
+void MLIRCompiler::build_ng_dialect_module()
+{
+    // initialize an empty module
+    m_module = make_unique<mlir::Module>(&m_context);
+
+    TypeList args_type_list, result_type_list;
+
+    // Retrieve input and output tensors.
+    const auto& kernel_inputs = m_compiled_kernel->get_arguments();
+    const auto& kernel_outputs = m_compiled_kernel->get_kernel_outputs();
+    NGRAPH_ASSERT(kernel_inputs.size() != 0) << "Cannot have empty inputs list";
```

Comment:
`NGRAPH_ASSERT` is deprecated. Please use `NGRAPH_CHECK` instead.

Reply:
Replaced. Thanks!